File: ios/Sources/Sketch/SketchScene.swift
import Foundation
import os
import RealityKit
import UIKit

/// THROWAWAY (SMASH-5): the clickable motion sketch — Title → Season hub → Match HUD → Result,
/// and the coach's board — over the Oasis world, built only from the UI kit, for the owner to
/// judge the 3D UI's feel on a phone (AGENTS: "Click-dummy new UI first"). Every transition is a
/// camera swoop with things tumbling out and in; nothing cuts.
@MainActor
final class SketchScene {
    let root = Entity()
    let stage: UIStage
    private let world: OasisWorld
    private let stats = FrameStats()
    private let log = Logger(subsystem: "in.nann.smashhockey", category: "sketch")
    private var screens: [SketchScreenID: SketchScreen] = [:]
    private(set) var current: SketchScreenID?
    private var title: TitleScreen!
    private var hub: HubScreen!
    private var match: MatchScreen!
    private var result: ResultScreen!
    private var coach: CoachScreen!

    private init(stage: UIStage, world: OasisWorld) {
        self.stage = stage
        self.world = world
        root.addChild(world.root)
        root.addChild(stage.root)
    }

    static func make(viewSize: CGSize, insets: UIEdgeInsets) async throws -> SketchScene {
        let motion = try MotionTokens.load()
        try TextMesh.registerFont()
        let start = CameraPose(eye: [0, 30, -70], target: [0, 4, 0])
        let world = try await OasisWorld.load(eye: start.eye)
        let stage = UIStage(pose: start, viewSize: viewSize, insets: insets, motion: motion)
        let scene = SketchScene(stage: stage, world: world)
        scene.build()
        return scene
    }

    private func build() {
        let go: Go = { [weak self] id in self?.go(to: id) }
        title = TitleScreen(stage: stage, go: go)
        hub = HubScreen(stage: stage, go: go)
        match = MatchScreen(stage: stage, world: world.root, go: go)
        result = ResultScreen(stage: stage, go: go)
        coach = CoachScreen(stage: stage, go: go)
        screens = [.title: title, .hub: hub, .match: match, .result: result, .coach: coach]
        match.onFinished = { [weak self] home, away in
            guard let self else { return }
            self.result.setScore(home, away)
            self.hub.applyMatchday()
            self.go(to: .result)
        }
        self.go(to: .title)
    }

    /// The one way between screens: the old one's parts hop away, the camera swoops, the new
    /// one's parts tumble in as it arrives.
    func go(to id: SketchScreenID) {
        guard id != current, let next = screens[id] else { return }
        log.info("screen \(id.rawValue, privacy: .public) t=\(self.stage.time, format: .fixed(precision: 2))")
        if let c = current { screens[c]?.hide() }
        let first = current == nil
        current = id
        stage.rig.swoop(to: next.pose, roll: first ? 0 : 0.12)
        next.show(after: first ? 0.9 : 0.45)
    }

    /// The one clock (ADR 0005): the scene's update event, once per rendered frame.
    func update(_ dt: Double) {
        stats.record(dt)
        let step = min(dt, 0.1)
        if let c = current { screens[c]?.update(step) }
        stage.update(step)
        world.setFogEye(stage.rig.pose.eye)
    }

    // MARK: autoplay (-sketchAutoplay)

    /// Walks the whole sketch on a timer, through the same actions a finger fires, so every
    /// screen and mid-transition frame can be captured from the simulator.
    func startAutoplay() {
        let steps: [(Double, @MainActor (SketchScene) -> Void)] = [
            (4.0, { $0.tap($0.title.training) }),
            (5.0, { $0.tap($0.title.play) }),
            (10.0, { $0.tap($0.hub.playMatch) }),
            (19.0, { $0.tap($0.match.pause) }),
            (22.0, { $0.tap($0.match.resume) }),
            // the match finishes on its own (~45 s) and swoops to the result
            (52.0, { $0.tap($0.result.next) }),
            (58.0, { $0.tap($0.hub.back) }),
            (62.0, { $0.tap($0.title.coach) }),
            (65.0, { $0.drag($0.coach.sliders[0], to: 0.85) }),
            (66.0, { $0.drag($0.coach.sliders[1], to: 0.2) }),
            (67.0, { $0.drag($0.coach.sliders[2], to: 0.7) }),
            (69.0, { $0.tap($0.coach.back) }),
        ]
        for (t, run) in steps { stage.after(t) { [weak self] in if let self { run(self) } } }
    }

    /// A pretend finger on the button's projected rectangle: the overlay's frame, fed back through
    /// the stage's own ray hit-test — so autoplay proves projection and picking agree.
    private func tap(_ b: BlockButton) {
        guard let p = centre(of: b.semantics.id) else { return }
        stage.touchDown(at: p)
        stage.after(0.14) { [weak self] in self?.stage.touchUp(at: p) }
    }

    /// A pretend drag along a slider, from its value to `value`, in eight moves.
    private func drag(_ s: Slider3D, to value: Double) {
        guard let r = stage.semanticsModel.nodes.first(where: { $0.id == s.semantics.id })?.frame else {
            log.error("autoplay: \(s.semantics.id, privacy: .public) is not on screen"); return
        }
        // The rail spans the node's box minus 0.1 m of grab room each side (Slider3D.bounds).
        let pad = r.width * 0.1 / CGFloat(s.length + 0.2)
        func x(_ v: Double) -> CGFloat { r.minX + pad + (r.width - 2 * pad) * v }
        let y = r.midY + r.height * 0.2
        stage.touchDown(at: CGPoint(x: x(s.value), y: y))
        let from = s.value
        for i in 1...8 {
            let v = from + (value - from) * Double(i) / 8
            stage.after(Double(i) * 0.05) { [weak self] in self?.stage.touchMoved(to: CGPoint(x: x(v), y: y)) }
        }
        stage.after(0.5) { [weak self] in self?.stage.touchUp(at: CGPoint(x: x(value), y: y)) }
    }

    private func centre(of id: String) -> CGPoint? {
        guard let r = stage.semanticsModel.nodes.first(where: { $0.id == id })?.frame else {
            log.error("autoplay: \(id, privacy: .public) is not on screen")
            return nil
        }
        return CGPoint(x: r.midX, y: r.midY)
    }
}

File: android/app/src/main/kotlin/in/nann/smashhockey/sketch/SketchScene.kt
package `in`.nann.smashhockey.sketch

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Typeface
import android.util.Log
import android.view.MotionEvent
import android.view.SurfaceView
import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.android.filament.Camera
import com.google.android.filament.ColorGrading
import com.google.android.filament.ToneMapper
import com.google.android.filament.View
import `in`.nann.smashhockey.core.generated.World as WorldId
import `in`.nann.smashhockey.engine.Assets
import `in`.nann.smashhockey.engine.FilamentHost
import `in`.nann.smashhockey.engine.FrameStats
import `in`.nann.smashhockey.engine.World
import `in`.nann.smashhockey.ui.BlockButton
import `in`.nann.smashhockey.ui.CameraPose
import `in`.nann.smashhockey.ui.Kit
import `in`.nann.smashhockey.ui.Motion
import `in`.nann.smashhockey.ui.Slider3D
import `in`.nann.smashhockey.ui.UIStage

/**
 * THROWAWAY (SMASH-5): the clickable motion sketch — the twin of iOS's SketchScene: Title → Season
 * hub → Match HUD → Result, and the coach's board — over the Oasis world, built only from the UI
 * kit, for the owner to judge the 3D UI's feel on a phone (AGENTS: "Click-dummy new UI first").
 * Every transition is a camera swoop with things tumbling out and in; nothing cuts.
 *
 * The screens are laid out from the surface's size and the safe area, so they are built on the
 * first frame that knows both.
 */
class SketchScene(
    context: Context,
    private val surfaceView: SurfaceView,
    private val autoplay: Boolean,
    private val forceReduceMotion: Boolean,
) {
    private val host = FilamentHost(surfaceView, ::frame)
    private val engine = host.engine
    private val assets = Assets(context.assets)
    private val stats = FrameStats(context)
    private val kit = Kit(engine, host.scene, assets, Typeface.createFromAsset(context.assets, "fonts/LilitaOne-Regular.ttf"),
        Motion.load(assets))
    private val world = World(engine, host.scene, host.view, assets, WorldId.OASIS)
    private val copy = Copy { context.getString(it) }
    private val stageState = mutableStateOf<UIStage?>(null)
    /** The stage, once built — the semantics overlay reads it. */
    val stage: State<UIStage?> get() = stageState

    private val screens = HashMap<SketchScreenId, SketchScreen>()
    private var current: SketchScreenId? = null
    private lateinit var title: TitleScreen
    private lateinit var hub: HubScreen
    private lateinit var match: SketchMatchScreen
    private lateinit var result: ResultScreen
    private lateinit var coach: CoachScreen
    private var width = 0
    private var height = 0
    private var touching = false

    init {
        host.view.colorGrading = ColorGrading.Builder().toneMapper(ToneMapper.Linear()).build(engine)
        host.view.dynamicResolutionOptions = View.DynamicResolutionOptions().apply { enabled = true; minScale = 0.6f }
        host.camera.setExposure(16f, 1f / 125f, 100f)
        host.onResize = { w, h ->
            width = w; height = h
            host.camera.setProjection(FOV, w.toDouble() / h, 0.1, 500.0, Camera.Fov.VERTICAL)
            stageState.value?.resize(w, h)
        }
    }

    private fun build() {
        val insets = ViewCompat.getRootWindowInsets(surfaceView)
            ?.getInsets(WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout())
        val start = CameraPose(0f, 30f, -70f, 0f, 4f, 0f)
        val stage = UIStage(kit, host.camera, host.camera.entity, start, width, height, insets?.top ?: 0, insets?.bottom ?: 0)
        stage.reduceMotion = reduceMotion()
        val go: Go = { go(it) }
        title = TitleScreen(stage, copy, go)
        hub = HubScreen(stage, copy, go)
        match = SketchMatchScreen(stage, copy, kit.node(null), go)
        result = ResultScreen(stage, copy, go)
        coach = CoachScreen(stage, copy, go)
        screens.putAll(listOf(SketchScreenId.TITLE to title, SketchScreenId.HUB to hub, SketchScreenId.MATCH to match,
            SketchScreenId.RESULT to result, SketchScreenId.COACH to coach))
        match.onFinished = { home, away ->
            result.setScore(home, away)
            hub.applyMatchday()
            go(SketchScreenId.RESULT)
        }
        stageState.value = stage
        go(SketchScreenId.TITLE)
        if (autoplay) startAutoplay(stage)
    }

    /** Reduce Motion: the system's animator duration scale at 0, or `--ez sketchReduceMotion true`. */
    private fun reduceMotion() = forceReduceMotion || !ValueAnimator.areAnimatorsEnabled()

    /**
     * The one way between screens: the old one's parts hop away, the camera swoops, the new one's
     * parts tumble in as it arrives.
     */
    fun go(id: SketchScreenId) {
        val stage = stageState.value ?: return
        if (id == current) return
        val next = screens[id] ?: return
        Log.i(TAG, "screen ${id.name.lowercase()} t=${"%.2f".format(stage.time)}")
        current?.let { screens[it]?.hide() }
        val first = current == null
        current = id
        stage.rig.swoop(next.pose, roll = if (first) 0f else 0.12f)
        next.show(if (first) 0.9 else 0.45)
    }

    fun start() = host.start()
    fun stop() = host.stop()

    /** The one clock (ADR 0005): the Choreographer frame. */
    private fun frame(dt: Double, @Suppress("UNUSED_PARAMETER") frameTimeNanos: Long) {
        stats.record(dt)
        if (stageState.value == null) {
            if (width <= 1 || height <= 1) return
            build()
        }
        val stage = stageState.value ?: return
        val step = minOf(dt, 0.1)
        stage.reduceMotion = reduceMotion()
        current?.let { screens[it]?.update(step) }
        stage.update(step)
    }

    /** Fingers reach the stage's own hit-test: the first finger only, like iOS's drag gesture. */
    fun onTouch(e: MotionEvent): Boolean {
        val stage = stageState.value ?: return true
        when (e.actionMasked) {
            MotionEvent.ACTION_DOWN -> { touching = true; stage.touchDown(e.x, e.y) }
            MotionEvent.ACTION_MOVE -> if (touching) stage.touchMoved(e.x, e.y)
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> if (touching) { touching = false; stage.touchUp(e.x, e.y) }
        }
        return true
    }

    // ------------------------------------------------------------------ autoplay (--ez sketchAutoplay true)

    /**
     * Walks the whole sketch on a timer, through the same actions a finger fires — the same
     * schedule as iOS's `-sketchAutoplay`.
     */
    private fun startAutoplay(stage: UIStage) {
        val steps: List<Pair<Double, () -> Unit>> = listOf(
            4.0 to { tap(stage, title.training) },
            5.0 to { tap(stage, title.play) },
            10.0 to { tap(stage, hub.playMatch) },
            19.0 to { tap(stage, match.pause) },
            22.0 to { tap(stage, match.resume) },
            // the match finishes on its own (~45 s) and swoops to the result
            52.0 to { tap(stage, result.next) },
            58.0 to { tap(stage, hub.back) },
            62.0 to { tap(stage, title.coach) },
            65.0 to { drag(stage, coach.sliders[0], 0.85) },
            66.0 to { drag(stage, coach.sliders[1], 0.2) },
            67.0 to { drag(stage, coach.sliders[2], 0.7) },
            69.0 to { tap(stage, coach.back) },
        )
        for ((t, run) in steps) stage.after(t, run)
    }

    /**
     * A pretend finger on the button's projected rectangle: the overlay's frame, fed back through
     * the stage's own ray hit-test — so autoplay proves projection and picking agree.
     */
    private fun tap(stage: UIStage, b: BlockButton) {
        val r = stage.rect(b.semantics.id) ?: run { Log.e(TAG, "autoplay: ${b.semantics.id} is not on screen"); return }
        val x = (r.left + r.right) / 2f; val y = (r.top + r.bottom) / 2f
        stage.touchDown(x, y)
        stage.after(0.14) { stage.touchUp(x, y) }
    }

    /** A pretend drag along a slider, from its value to [value], in eight moves. */
    private fun drag(stage: UIStage, s: Slider3D, value: Double) {
        val r = stage.rect(s.semantics.id) ?: run { Log.e(TAG, "autoplay: ${s.semantics.id} is not on screen"); return }
        val w = (r.right - r.left).toFloat()
        // The rail spans the node's box minus 0.1 m of grab room each side (Slider3D.bounds).
        val pad = w * 0.1f / (s.length + 0.2f)
        fun x(v: Double) = r.left + pad + (w - 2 * pad) * v.toFloat()
        val y = (r.top + r.bottom) / 2f + (r.bottom - r.top) * 0.2f
        stage.touchDown(x(s.value), y)
        val from = s.value
        for (i in 1..8) {
            val v = from + (value - from) * i / 8
            stage.after(i * 0.05) { stage.touchMoved(x(v), y) }
        }
        stage.after(0.5) { stage.touchUp(x(value), y) }
    }

    fun destroy() {
        host.stop()
        kit.destroy()
        world.destroy()
        host.destroy()
    }

    companion object {
        const val TAG = "SmashSketch"
        private const val FOV = 50.0
    }
}
